### PR TITLE
fix: various typescript errors and better performance of recursive helpers

### PR DESCRIPTION
**Pull request by grabbou**
Fixing some annoying TypeScript errors in order to make it all work nicely locally and for users. The project builds fine at this stage.

Fixes explained line by line.

### CODE REVIEW

File: patches/@tamagui+progress+1.94.4.patch [hunk 1]
```diff
@@ -0,0 +1,11 @@
+diff --git a/node_modules/@tamagui/progress/types/Progress.d.ts b/node_modules/@tamagui/progress/types/Progress.d.ts
+index 58f3db4..cbdfac4 100644
+--- a/node_modules/@tamagui/progress/types/Progress.d.ts
++++ b/node_modules/@tamagui/progress/types/Progress.d.ts
+@@ -254,5 +254,5 @@ declare const Progress: React.ForwardRefExoticComponent<Omit<import("@tamagui/co
+     }, import("@tamagui/core").StaticConfigPublic>;
+ };
+ export { createProgressScope, Progress, ProgressIndicator };
+-export type { ProgressProps, ProgressIndicatorProps };
++export type { ProgressProps, ProgressExtraProps, ProgressIndicatorProps };
```
Comment (grabbou) [inline patches/@tamagui+progress+1.94.4.patch:10]: Temporarily using `patch-package` but hoping to send PR to Tamagui. W/o "ExtraProps" exported, it is not possible to re-export Tamagui-derived type with "declaration: true".

Issue that is encountered: https://github.com/microsoft/TypeScript/issues/5711

File: packages/playground/src/data-boundary.tsx [hunk 2]
```diff
@@ -9,8 +9,8 @@ export function DataBoundary({
   dataSource,
   path,
   children,
-}: PropsWithChildren<{ dataSource: WebSocketDataSource; path: string }>) {
-  const state = useStream(dataSource.state)
+}: PropsWithChildren<{ dataSource: WebSocketDataSource | DataSource; path: string }>) {
+  const state = useStream('state' in dataSource ? dataSource.state : undefined)
```
Comment (grabbou) [inline packages/playground/src/data-boundary.tsx:13]: fix: typescript error when passing HTTPDataSource to Boundary.

The extra conditional is just to make TypeScript happy. `useStream` handles undefined already. That's why previous code was working just fine.

Unfortunately, `dataSource?.state` is going to fail too.

File: packages/react/src/jsx.tsx [hunk 3]
```diff
@@ -13,27 +12,20 @@ import {
...
-  | ActionDataState
-  | ReferencedDataState
-  | StateDataState
-  | ((args: any) => any)
+export type UI = ReactElement<Props> | ServerComponent
 
-type Props = Record<string, JSONValue | AllowedDataStates> & {
-  children: DataState
+type ServerComponent = ComponentDataState<ServerEventDataState>
+type Props = ServerComponent['props'] & {
```
Comment (grabbou) [inline packages/react/src/jsx.tsx:18]: refactor: do not duplicate "prop" types that can be defined for components on the server. Instead, reference props from the real type.

File: packages/react/src/template.tsx [hunk 4]
```diff
@@ -13,28 +13,27 @@ export type ComponentDefinition<T extends Record<string, JSONValue>> = {
 }
 
 /** Data state */
-export type DataState<T = never> =
+export type DataState<T = EventDataState> =
```
Comment (grabbou) [inline packages/react/src/template.tsx:16]: fix: ServerDataState (DataState on the server, before serialisation) now will not have EventDataState. This is correct. EventDataState without function on the server is going to result in an error, so should be properly typed. I will improve this type in a follow-up anyway.

File: packages/react/src/template.tsx [hunk 5]
```diff
@@ -13,28 +13,27 @@ export type ComponentDefinition<T extends Record<string, JSONValue>> = {
...
-export type ComponentDataState<T = never> = {
+export type ComponentDataState<T = EventDataState> = {
   $: 'component'
   key?: string
   component: ComponentIdentifier
   children?: DataState<T>
   props?: Record<string, DataState<T>>
 }
-type ReferencedComponentDataState<T = never> = ComponentDataState<T> & {
+type ReferencedComponentDataState<T = undefined> = ComponentDataState<T> & {
```
Comment (grabbou) [inline packages/react/src/template.tsx:36]: fix: by turning this into `undefined`, we can skip <T> when we want the default.

File: packages/react/src/utils.ts [hunk 6]
```diff
@@ -1,4 +1,4 @@
-export function lookupValue(value: Record<string, any>, ref: (string | number)[]) {
+export function lookupValue(value: unknown, ref: (string | number)[]) {
```
Comment (grabbou) [inline packages/react/src/utils.ts:1]: fix: we accept any value and we check for different types inside (we also have unit tests for it)

File: packages/ws-client/src/data-source.ts [hunk 7]
```diff
@@ -34,7 +34,7 @@ export type UpdateWebsocketMessage = {
 export type EventResponseWebsocketMessage = {
   $: 'evt-res'
   key: string
-  res: ServerResponseDataState
+  res: ResponseDataState
```
Comment (grabbou) [inline packages/ws-client/src/data-source.ts:37]: fix: the type on the client (after json.stringify) is obv. ResponseDataState (object w/o functions)

File: packages/react/src/utils.ts (hunk 6 above)
Comment (ericvicenti) [reply packages/react/src/utils.ts:1]: ok great, relying on inference here

File: packages/react/src/template.tsx [hunk 8]
```diff
@@ -13,28 +13,27 @@ export type ComponentDefinition<T extends Record<string, JSONValue>> = {
...
-  | EventDataState
   | StateDataState
   | { [key: string]: DataState<T>; $?: never }
   | DataState[]
   | T
 /** Server data state */
 export type ServerDataState = DataState<ServerEventDataState>
 
-export type ComponentDataState<T = never> = {
+export type ComponentDataState<T = EventDataState> = {
```
Comment (ericvicenti) [inline packages/react/src/template.tsx:29]: do we want to use `extends EventDataState` here as well?

File: packages/react/src/template.tsx (hunk 5 above)
Comment (ericvicenti) [reply packages/react/src/template.tsx:36]: hmm, when is the referenced value supposed to be undefined? Not sure I understand this part

File: packages/react/src/jsx.tsx [hunk 9]
```diff
@@ -58,15 +58,24 @@ export function jsx(
   }
 }
 
-type Extend<P, K> = {
-  [Key in keyof P]: P[Key] extends infer T ? (T extends object ? K | Extend<T, K> : T | K) : never
-}
```
[the remark this answers is not in the document]
Comment (ericvicenti) [reply packages/react/src/jsx.tsx:63]: Ah man I used to run into this all the time on zerve when I used super heavy TS inference

**Review by ericvicenti — APPROVED** [review]
Man, library TS is tricky, I barely understand 😂

Bummer we have to use patch-package especially for a peerDep which is scary, but hopefully it doesn't last long

File: packages/react/src/template.tsx (hunk 5 above)
Comment (grabbou) [reply packages/react/src/template.tsx:36]: LOL I thought this was working but it's obviously not. I wanted to make `ReferencedComponentDataState` completely optional but accidentally ended up passing `undefined` to ComponentDataState. Fixed

**Comment by grabbou** [comment]
Yeah @ericvicenti I already spoke to Nate about this and hopefully he ships a fix soon!

**Pull request merged**